cmd/iam_role_create.go

```diff
@@ -62,71 +62,80 @@ func (c *iamRoleCreateCmd) cmdRun(cmd *cobra.Command, _ []string) error {
 		exoapi.NewReqEndpoint(account.CurrentAccount.Environment, zone),
 	)
 
-	if c.Policy == "-" {
-		inputReader := cmd.InOrStdin()
-		b, err := io.ReadAll(inputReader)
-		if err != nil {
-			return fmt.Errorf("failed to read policy from stdin: %w", err)
-		}
+	var policy *exoscale.IAMPolicy
 
-		c.Policy = string(b)
-	}
-
-	var obj iamPolicyOutput
-	err := json.Unmarshal([]byte(c.Policy), &obj)
-	if err != nil {
-		return fmt.Errorf("failed to parse policy: %w", err)
-	}
+	// Policy is optional, if not set API will default to `allow all`
+	if c.Policy != "" {
+		// If Policy value is `-` read from STDIN
+		if c.Policy == "-" {
+			inputReader := cmd.InOrStdin()
+			b, err := io.ReadAll(inputReader)
+			if err != nil {
+				return fmt.Errorf("failed to read policy from stdin: %w", err)
+			}
 
-	policy := &exoscale.IAMPolicy{
-		DefaultServiceStrategy: obj.DefaultServiceStrategy,
-		Services:               map[string]exoscale.IAMPolicyService{},
-	}
+			c.Policy = string(b)
+		}
 
-	if len(obj.Services) > 0 {
-		for name, sv := range obj.Services {
-			service := exoscale.IAMPolicyService{
-				Type: func() *string {
-					t := sv.Type
-					return &t
-				}(),
-			}
+		var obj iamPolicyOutput
+		err := json.Unmarshal([]byte(c.Policy), &obj)
+		if err != nil {
+			return fmt.Errorf("failed to parse policy: %w", err)
+		}
 
-			if len(sv.Rules) > 0 {
-				service.Rules = []exoscale.IAMPolicyServiceRule{}
-				for _, rl := range sv.Rules {
+		policy = &exoscale.IAMPolicy{
+			DefaultServiceStrategy: obj.DefaultServiceStrategy,
+			Services:               map[string]exoscale.IAMPolicyService{},
+		}
 
-					rule := exoscale.IAMPolicyServiceRule{
-						Action: func() *string {
-							t := rl.Action
-							return &t
-						}(),
-					}
+		if len(obj.Services) > 0 {
+			for name, sv := range obj.Services {
+				service := exoscale.IAMPolicyService{
+					Type: func() *string {
+						t := sv.Type
+						return &t
+					}(),
+				}
 
-					if rl.Expression != "" {
-						rule.Expression = func() *string {
-							t := rl.Expression
-							return &t
-						}()
+				if len(sv.Rules) > 0 {
+					service.Rules = []exoscale.IAMPolicyServiceRule{}
+					for _, rl := range sv.Rules {
+
+						rule := exoscale.IAMPolicyServiceRule{
+							Action: func() *string {
+								t := rl.Action
+								return &t
+							}(),
+						}
+
+						if rl.Expression != "" {
+							rule.Expression = func() *string {
+								t := rl.Expression
+								return &t
+							}()
+						}
+
+						service.Rules = append(service.Rules, rule)
 					}
-
-					service.Rules = append(service.Rules, rule)
 				}
-			}
 
-			policy.Services[name] = service
+				policy.Services[name] = service
+			}
 		}
 	}
```

Comment:
I count 6 layers of nesting :)
Maybe you could split up some of this functionality into smaller functions.

Reply:
I just re-arranged what was already there, i do agree code is difficult to traverse. I'll look into refactoring.

Reply:
Fixed